SI-5361 Avoid cyclic type with malformed refinement

retronym opened the pull request:
The statement `val x = this` in the refinment type:

```
(new {}): {val x = this}
```

is lazily typechecked, in order to, according to the comment
in `typedRefinment, "avoid cyclic reference errors".

But the approximate type used ends up with:

```
Refinment@1(
  parents = [...]
  decls = { val x: Refinement@1 })
```

This commit eagerly checks that there is no term definitions
in type refinments, rather than delaying this.

This changes the error message for SI-3614.

Review by @adriaanm

retronym committed:
SI-5361 Avoid cyclic type with malformed refinement

The statement `val x = this` in the refinment type:

    (new {}): {val x = this}

is lazily typechecked, in order to, according to the comment
in `typedRefinment, "avoid cyclic reference errors".

But the approximate type used ends up with:

    Refinment@1(
      parents = [...]
      decls = { val x: Refinement@1 })

This commit eagerly checks that there is no term definitions
in type refinments, rather than delaying this.

This changes the error message for SI-3614.

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5068,7 +5074,7 @@ trait Typers extends Modes with Adaptations with Tags {
           newTyper(context.make(templ, self.typeSymbol, decls)).typedRefinement(templ)
           templ updateAttachment CompoundTypeTreeOriginalAttachment(parents1, Nil) // stats are set elsewhere
-          tree setType self
+          tree setType (if (templ.exists(_.isErroneous)) ErrorType else self) // Being conservative to avoid SI-5361
```
How about this check is put a couple of lines above, along its friend (parents1 exists (_.isErrorTyped)) ?

retronym replied:
We rely on `typedRefinement` having been called. Alternatively, I could pull the `OnlyDeclarationsError` out of there and into `typedCompoundTypeTree`. Not sure which way to go here; wdyt?

hubertp replied:
Ah, right. Yeah, I think it might be a bit better.
I really don't like that typedRefinement returns Unit (unlike other typedXYZ) .

paulp replied:
paulp says: all Unit-returning methods which aren't println/logging are bugs.

retronym committed:
SI-5361 Refactor in accordance with review comments.

Check for malformed stats in the refinement body in
the same place we check for erroneous parents.

retronym commented:
@hubertp I've shuffled the check around.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1792/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1792/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1080/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1080/

adriaanm commented:
LGTM
